src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs

```diff
@@ -0,0 +1,16 @@
+// run-pass
+
+#![feature(track_caller)]
+#![allow(dead_code)]
+
+trait Trait {
+    #[track_caller] //~ WARN: `#[track_caller]` is ignored on function prototypes
+    fn foo(&self);
+}
+
+extern "Rust" {
+    #[track_caller] //~ WARN: `#[track_caller]` is ignored on function prototypes
+    fn bar();
```

Comment:
Btw we might want this to actually work and change the ABI to pass the caller!

Reply:
Would that still apply if the ABI is e.g. `"C"`?

Reply:
I think we'd want that to be an error (it should already be in general, e.g. if you're *defining* an `extern "C" fn`, i.e. exporting it).

Reply:
We do have an error without the Rust ABI when exporting:
```rust
#![feature(track_caller)]

#[track_caller]
extern "C" fn foo() {} //~ ERROR `#[track_caller]` requires Rust ABI
```

@anp I don't recall seeing anything about this in the reference PR (https://github.com/rust-lang/reference/pull/742, let's make sure we do note that restriction). Also, do we have a test checking this? (E.g. for "C" and some other random obscure ABI.)

In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

Reply:
> In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

This PR changes that to a warning (which is what we're commenting on).

Reply:
Yeah, let's promote to an error, cover it in the reference, and punt on ABI semantics questions?

Reply:
Sounds good. We do have a test `error-with-invalid-abi.rs` :)

Reply:
https://github.com/rust-lang/reference/pull/742/files#r396035394 for reference docs, changing `warn-fn-prototype.rs` -> `error-extern-fn.rs`.